feat: Advanced Setting for Dev Mode Local Telemetry Logging

klewis-sfdc opened the pull request:
<!--- PR title should follow the pattern: <type>(optional scope): <description>.
please refer to the types and format here: https://www.conventionalcommits.org/en/v1.0.0/#summary
If this is a feat/fix, add the technical writer as a reviewer to the PR. --->

### What does this PR do?
* When the Telemetry Service instance is initialized, it will check the settings.json file for an entry for `<<extension-name>>.advanced.localTelemetryLogging`.   If set to 'true', when running in Debug Mode, the named extension(s) will log Telemetry events to a local "<<extensionName>>-telemetry.json" file.  This is intended to be helpful when validating telemetry while developing and debugging features.

### What issues does this PR fix or reference?
@W-14983101@

### Functionality Before
* Telemetry can be streamed to a file on the file system when in Production Mode (VSIX):
  * Add the following Env variables in a terminal (I ran, for example, "export VSCODE_LOGS=/dev/LOG" in my terminal to set the first env variable)
    * VSCODE_LOGS=path/to/where/you/want/log/file
    * VSCODE_LOG_LEVEL=trace 
  * Then launch VS Code from that same terminal to pick up the env vars by running "code ."
    * run "env" in a terminal to list the env variables that are present.

### Functionality After
* The above still works in production mode, but Telemetry can now also be appended to a 'telemetry.json' file at the project root when in Development Mode (launched from IDE):

Add:     
>     "<<extension-name>>.advanced": {
>         "localTelemetryLogging": "true"
>     }

to your **settings.json** file, i.e.:

```
    "salesforcedx-einstein-gpt.advanced": {
        "localTelemetryLogging": "true"
    },
    "salesforcedx-vscode-core.advanced": {
        "localTelemetryLogging": "true"
    }
```

Note: most extensions currently use the telemetry instance from the core extension, so enabling and disabling logging this way
```
    "salesforcedx-vscode-core.advanced": {
        "localTelemetryLogging": "true"
    }
```
...will give a lot of telemetry events for the extensions that are part of the extension pack. 

Questions:

- [ ] Should we standardized on one of the local loggers?
  -  I kind of like each of these loggers for different reasons.  I think they each have their purpose.  I wrote up some of the details here: https://salesforce.quip.com/VNd9AkciwI5k

TODOs:

- [x] Unit Tests for TelemetryFile logger
- [x] WI for this work
- [-] ~~Make the TelemetryFile logger use fs.writeStream~~
- [ ] Follow up WI to make each extension use its own instance of TelemetryService (TBD)
- [x] Pare down the console.logging
- [ ] Advanced setting set to **false** should turn off local TelemFile logging
- [ ] update developing.md

gbockus-sf commented on packages/salesforcedx-utils-vscode/src/telemetry/localTelemetryReporter.ts:
```diff
@@ -0,0 +1,27 @@
+  ) {
+    const timestamp = new Date().toISOString();
+    appendFileSync(
+      'telemetry.json',
```
Nit: make this a constant in the file and maybe call it E4DTelemetry.json (or something more specific)

gbockus-sf commented on packages/salesforcedx-utils-vscode/src/telemetry/localTelemetryReporter.ts:
```diff
@@ -0,0 +1,27 @@
+    }
+  ) {
+    const timestamp = new Date().toISOString();
+    appendFileSync(
```
Can this be async?

gbockus-sf commented on packages/salesforcedx-utils-vscode/src/settings/sfdxSettingsService.ts:
```diff
@@ -5,12 +5,26 @@
+  ): boolean {
+    const isLocalTelemetryLoggingEnabled = vscode.workspace
+      .getConfiguration()
+      .get<string>(`${extensionName}.advanced.enableLocalTelemetryLogging`);
```
I wish this could be a constant.  The right half could be anyway.

gbockus-sf commented on packages/salesforcedx-utils-vscode/src/settings/sfdxSettingsService.ts:
```diff
@@ -5,12 +5,26 @@
+      .getConfiguration()
+      .get<string>(`${extensionName}.advanced.enableLocalTelemetryLogging`);
+    // if the setting doesn't exist for this extension, false will be returned
+    const booleanVal = isLocalTelemetryLoggingEnabled === 'true';
```
constant for 'true'

gbockus-sf commented on packages/salesforcedx-utils-vscode/src/telemetry/telemetry.ts:
```diff
@@ -139,6 +140,17 @@ export class TelemetryService {
+        SfdxSettingsService.isLocalTelemetryLoggingEnabledFor(
+          this.extensionName
+        );
+
```
can we add a logger here

klewis-sfdc replied:
this repo needs a logger :)

gbockus-sf commented on packages/salesforcedx-utils-vscode/src/telemetry/telemetry.ts:
```diff
@@ -238,12 +257,28 @@ export class TelemetryService {
+    if (this.localReporter) {
+      this.localReporter.writeToFile(eventName, {
+        ...properties
+      });
```
Should we try/catch and only log in the catch around this?  having actual errors thrown out of this service due to local logging would be a bummer

gbockus-sf reviewed (approved):
I like it...some comments for your consideration

gbockus-sf reviewed (commented):
I meant to comment but approved like a doofus.

gbockus-sf reviewed (changes requested):
comments inline...accidental approval :P

peternhale commented on packages/salesforcedx-utils-vscode/src/telemetry/telemetry.ts:
I appreciate the ability to to direct telemetry to local file system, but adding a distinct local reporter and the needed changes within the telemetry to send data to each seems messy to me. Would the changing the representation of reporter to `private reporter: TelemetryReporter[]` and then have TelemetryService dispatch each `send` call to each reporter make better sense?

klewis-sfdc replied:
Yeah, I like that idea!  I'll see what that looks like

klewis-sfdc commented on packages/salesforcedx-utils-vscode/src/commands/channelService.ts:
```diff
@@ -9,7 +9,7 @@ import stripAnsi from 'strip-ansi';
 import { CommandExecution } from '../cli';
 import { nls } from '../messages';
-import { SfdxSettingsService } from '../settings';
+import { SettingsService } from '../settings';
```
This rename touches a lot of files, but felt necessary.

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/settings/advancedSettings.ts:
```diff
@@ -0,0 +1,9 @@
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+export enum AdvancedSettings {
+  ENABLE_LOCAL_TELEMETRY_LOGGING = 'enableLocalTelemetryLogging'
```
if only developers are setting this, we could shorten the name. Setting to `true` could help remove ambiguity in the name too, e.g. `localTelemetryLogging:true`=> implies it's enabled

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/settings/settingsService.ts:
```diff
@@ -0,0 +1,34 @@
+    extensionName: string,
+    advancedSetting: AdvancedSettings
+  ): boolean {
+    const isAdvancedSettingEnabled = vscode.workspace
+      .getConfiguration()
+      .get<string>(`${extensionName}.${ADVANCED}.${advancedSetting}`);
+    // if the setting doesn't exist for this extension, false will be returned
+    const booleanVal = isAdvancedSettingEnabled === TRUE;
+    return booleanVal;
```
```suggestion
return vscode.workspace
      .getConfiguration()
      .get<string>(
        `${extensionName}.${ADVANCED}.${ENABLE_LOCAL_TELEMETRY_LOGGING}`
      ) === TRUE;
```

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/telemetry/interfaces/telemetryReporter.ts:
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable header/header */
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
```
why this copyright? if it's a MS file/type, can we import it?

klewis-sfdc replied:
I kept this on the appInsights reporter and on the interface, since I extracted it from the original TelemetryReporter that had this copyright.

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/telemetry/reporters/appInsights.ts:
```diff
@@ -221,6 +186,6 @@ export class TelemetryReporter extends Disposable {
       }
     });
-    return Promise.all([flushEventsToAI, flushEventsToLogger]);
+    return Promise.all([flushEventsToAI]);
```
```suggestion
    return flushEventsToAI;
```

can reduce the method to return inline

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/settings/settingsService.ts:
```diff
@@ -0,0 +1,34 @@
+  SFDX_CORE_CONFIGURATION_NAME,
+  TRUE
+} from '../constants';
+import { AdvancedSettings } from './AdvancedSettings';
```
`./A` -> `./a`

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/telemetry/reporters/logStream.ts:
```diff
@@ -0,0 +1,89 @@
+/* eslint-disable header/header */
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ *--------------------------------------------------------*/
+
+'use strict';
```
why disable the header/copyright?

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/telemetry/reporters/logStream.ts:
```diff
@@ -0,0 +1,89 @@
+      this.logStream.end();
+    });
+
+    return Promise.all([flushEventsToLogger]);
```
same Promise comment as above

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/telemetry/reporters/logStreamConfig.ts:
```diff
@@ -0,0 +1,18 @@
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+export class LogStreamConfig {
+  public static logFilePath() {
```
NIT: add explicit return types

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/telemetry/reporters/logStreamConfig.ts:
```diff
@@ -0,0 +1,18 @@
+ */
+export class LogStreamConfig {
+  public static logFilePath() {
+    return process.env['VSCODE_LOGS'] || '';
```
use `??` to avoid falsy value

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/telemetry/reporters/telemetryFile.ts:
```diff
@@ -0,0 +1,64 @@
+import { TelemetryReporter } from '../interfaces';
+
+export class TelemetryFile implements TelemetryReporter {
+  appInsightsClient: undefined;
+  userOptIn: undefined;
+  logStream: undefined;
+
```
what're these used for? I only see them used in tests, but ideally they'd have types 😉

WillieRuemmele commented on packages/salesforcedx-utils-vscode/src/telemetry/reporters/telemetryFile.ts:
```diff
@@ -0,0 +1,64 @@
+
+  sendTelemetryEvent(
+    eventName: string,
+    properties?: { [key: string]: string } | undefined,
```
optional types are `T | undefined` by default, can omit this `| undefined` if you want

peternhale commented on packages/salesforcedx-utils-vscode/src/telemetry/reporters/telemetryFile.ts:
```diff
@@ -0,0 +1,60 @@
+    }
+  ) {
+    const timestamp = new Date().toISOString();
+    appendFile(
```
Could the promise version of appendFile be used here and then call the function with `void appendFile...` This is a fire and forget call anyways.

peternhale reviewed (commented):
Just a question

randi274 reviewed (commented):
Some additional comments/review

klewis-sfdc commented on packages/salesforcedx-utils-vscode/src/services/telemetry.ts:
```diff
@@ -231,18 +281,34 @@ export class TelemetryService {
+        try {
+          reporter.sendExceptionEvent(name, message);
+        } catch (error) {
+          console.log(
```
The t/c is here in the service to catch in case any of the reporters has a problem trying to record telem on the exception.  This t/c is inside the for-each so it should already be catching in this scenario for all reporters here.

klewis-sfdc commented on packages/salesforcedx-utils-vscode/src/services/telemetry.ts:
```diff
@@ -278,7 +346,7 @@ export class TelemetryService {
    */
   private validateTelemetry(callback: () => void): void {
-    if (this.reporter !== undefined) {
+    if (this.reporters.length > 0) {
```
Valid means there is _at least_ 1 reporter, and telemetry is enabled in User config.

It used to mean **there is 1 reporter** and telem enabled in User config.  

I think "validate" is still ok here, though for a caller of this method it means more like, "should I try to log telemetry?"

klewis-sfdc commented on packages/salesforcedx-utils-vscode/src/settings/settingsService.ts:
```diff
@@ -5,12 +5,32 @@
       .getConfiguration(SFDX_CORE_CONFIGURATION_NAME)
       .get<boolean>(SETTING_CLEAR_OUTPUT_TAB, false);
   }
+
```
The functionality or the setting value method here?  I don't know.  If not maybe we can have a follow up WI

klewis-sfdc commented on packages/salesforcedx-utils-vscode/src/telemetry/reporters/logStream.ts:
This LogStream is extracted from functionality that was inside of the telemetry module before, but was hard to discover.  Hoping that with the LogStream as a separate module it is easier to understand and test.

randi274 reviewed (approved):
I'm good on Qs! Let me know if you want help with QA, or if @gbockus-sf has got ya covered on it.

klewis-sfdc commented:
> I'm good on Qs! Let me know if you want help with QA, or if @gbockus-sf has got ya covered on it.

@randi274 if you have time and could start QA'ing this this week and keep @gbockus-sf in the loop, that would be awesome. Maybe we can check in at 🔋 ⏳ in the ⏰ to confirm the plan.  Thank you!

randi274 reviewed (changes requested):
After reviewing as a group today, can we get a docs update to go along with this PR? Maybe a new section in `[developing.md](https://github.com/forcedotcom/salesforcedx-vscode/blob/develop/contributing/developing.md)`, like a Local Development header with a subheader for Telemetry?

klewis-sfdc commented:
> After reviewing as a group today, can we get a docs update to go along with this PR? Maybe a new section in [developing.md](https://github.com/forcedotcom/salesforcedx-vscode/blob/develop/contributing/developing.md), like a Local Development header with a subheader for Telemetry?

Great idea @randi274.  I updated the developing.md doc to include a section for local telemetry logging and added some unit tests for the SettingService file. Let me know what you think.

randi274 commented on contributing/developing.md:
```diff
@@ -215,3 +220,29 @@ The npmrc allows for project-level [configuration](https://docs.npmjs.com/cli/v8
+
+With the above configuration, all extensions that re-use the core extension's telemetry module will log telemetry events to a local file at the project root.
+
+Extensions that are dependent on the core extension and use its telemetry module can also take advantage of this local logging functionality. For instance, the Einstein for Developers extension uses core telemetry but initializes its own instance. So with that extension installed the following line in settings.json will log telemetry events to a local file for Einstein for Developers only:
```
I feel like this might be better to put in the E4D extension, since we don't do as much looking for docs over here. What do you think?

randi274 commented on contributing/developing.md:
```diff
@@ -215,3 +220,29 @@ The npmrc allows for project-level [configuration](https://docs.npmjs.com/cli/v8
 
 Our nvmrc specifies the minimum node version required to run the project.
+
+### Development Mode Local Telemetry Logging
```
Yay! this is super helpful to have in here 👏

randi274 reviewed (approved):
🎉 🎉 🎉 

QE here: https://salesforce.quip.com/TT1bAZhR1603

klewis-sfdc commented:
@gbockus-sf Let us know when/if you are ready to approve this one and we'll get it merged thanks!
